### PR TITLE
[Fix] Advanced digitizing distance when snapping

### LINKED ISSUE #46128: Advanced digitizing panel options do not work at intersections
### What is the bug or the crash?

Additional digitizing tools do not work at intersections.
screenshot-2021-11-19 204652
in version 3.18 it works fine

### Steps to reproduce the issue

Additional digitizing tools do not work at intersections

### Versions

3.22

### Supported QGIS version

- [x] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

_No response_

### LINKED ISSUE #46352: Advanced digitizing works wrong with Snapping
### What is the bug or the crash?

bug

### Steps to reproduce the issue

1. Start edit layer in some projection (example Pulkovo 1995 GK Zone 11N) with advanced digitizing enabled.
2. Try to "snap" to point in another projection (for example: WGS-1984)
3. advanced digitizing will show and use ("copy data") WGS_1984 coordinates Instead of Pulkovo 1995 GK Zone 11N.
 3.1 It'is ok when you just "create one vertex" in feature, although "advanced digitizing" still show "WGS 1984" coordinates
 3.2 It's bug when you start to change **d, a** parametres in "advanced digitizing bar" - it creates blueprint with wrong coordinates.
4. it is imposible to work and create any right features cause for example WGS x,y = 115, 57 not even acceptable in Pulkovo 1995 GK Zone 11N.

Reason: **QGIS “on the fly” CRS transformation** doesn't work in 
snapping -> click -> *there should be CRS transformation but here just x,y copying" -> advanced digitizing

Full Video with bug: https://youtu.be/VIMkMosS1QY

https://user-images.githubusercontent.com/64549817/144590088-3da8e449-26ca-444e-9915-82e6ecbe72c6.mp4

### Versions

3.20, 3.22 possible almost all 3+

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Koyaani**
## Description

This PR fixes #46352. The destination CRS wasn't specified. So, when it is different from the layer, the `QgsGeometryUtils::closestPoint` moves the snapped point to the wrong location.

This PR also fixes #46128. The intersection snapping create a `mSnapMatch` object linked to no layer. Thus, when the `interpolatedPoint` method is called, it returns a null point and so, `point` takes the coordinates (0, 0).

### CODE REVIEW

**Comment by gioman** [comment]
@Koyaani does it also fixes https://github.com/qgis/QGIS/issues/46128 ?

**Comment by Koyaani** [comment]
@gioman No, it does not solve this issue :confused:

**Comment by lbartoletti** [comment]
@Koyaani Thanks. Can you add tests please ? ;)

**Comment by Koyaani** [comment]
@lbartoletti I am not totally sure that the new condition isn't too restrictive. Do you have an idea?
(And I will add tests :smiley:)

**Comment by Koyaani** [comment]
The latest commits add tests for the `QgsAdvancedDigitizingDockWidget`. There are tests on the constraints in the `testqgscadutils.cpp` file. But not everything can be tested from `QgsCadUtils` (like Z/M constraints or bugs in this issue). So I started to make a test file for the dock. It mixes core, app and gui... It is difficult to separate the 3.

I couldn't do any test on this dock without making some modifications. I modified the constructor of `QgsMapToolAddFeature` to add a specific `QgsAdvancedDigitizingDockWidget` as this is in the parent class `QgsMapToolDigitizeFeature`. Furthermore, I changed `cadCanvasMoveEvent` to `canvasMoveEvent` in the methods of `TestQgsMapToolAdvancedDigitizingUtils`. The `canvasMoveEvent` method wraps the `cadCanvasMoveEvent` method but applies the advanced digitizing constraints before.

This last change makes crash a test in `test_app_maptooladdfeaturelinez`, I'm in the process of fixing this issue.

I also made 2 additional changes: removing `context.mConstraint` and `context.zConstraint` because they were defined but not used. And the correction of a small problem. The distance field could be modified before a first vertex was digitized.

**Comment by nyalldawson** [comment]
This is still marked as a draft. Is that correct?

**Comment by Koyaani** [comment]
@nyalldawson Yes, adding tests to reproduce these bugs is much harder than I think :/

File: python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in [hunk 1]
```diff
@@ -538,6 +538,12 @@ Convenient method to get the M value from the line edit wiget
 .. versionadded:: 3.22
 %End
 
+    CadCapacities capacities() const;
+%Docstring
+Returns the capacities
+
+.. versionadded:: 3.24
```
Comment (lbartoletti) [inline python/gui/auto_generated/qgsadvanceddigitizingdockwidget.sip.in:545]:
```suggestion
.. versionadded:: 3.26
```

3.24 is now the LTR and unfortunately, it's not acceptable, for me, to back port this.

File: python/core/auto_generated/qgscadutils.sip.in [hunk 2]
```diff
@@ -87,10 +87,6 @@ Defines constraints for the :py:func:`QgsCadUtils.alignMapPoint()` method.
 
         QgsCadUtils::AlignMapPointConstraint xConstraint;
         QgsCadUtils::AlignMapPointConstraint yConstraint;
-
-        QgsCadUtils::AlignMapPointConstraint zConstraint;
```
Comment (lbartoletti) [inline python/core/auto_generated/qgscadutils.sip.in:91]: Not used today, but maybe it's worth to keep it. I don't know.

**Review by lbartoletti — APPROVED** [review]
Except minor review. LGTM

File: python/core/auto_generated/qgscadutils.sip.in (hunk 2 above)
Comment (Koyaani) [reply python/core/auto_generated/qgscadutils.sip.in]: Yes, I removed the commit.

**Comment by antoniolocandro** [comment]
Just wondering if this will be backported to 3.24 and the LTR

**Comment by justinbb** [comment]
Please backport to 3.22 LTR. An LTR with non-functional advanced digitizing is inconvenient.

**Pull request merged**